core/src/main/java/com/linecorp/armeria/client/endpoint/RampingUpWeightedRoundRobinStrategy.java

```diff
@@ -0,0 +1,433 @@
+            assert oldEndpoints != null;
+            for (final Iterator<Endpoint> i = oldEndpoints.iterator(); i.hasNext();) {
+                final Endpoint oldEndpoint = i.next();
+                for (Endpoint newEndpoint : newEndpoints) {
+                    if (oldEndpoint.equals(newEndpoint)) {
+                        if (oldEndpoint.weight() > newEndpoint.weight()) {
+                            i.remove();
+                            // The weight of the new endpoint is lower than old one so we just replace it.
+                            replacedOldEndpoints.add(newEndpoint);
+                        } else if (oldEndpoint.weight() < newEndpoint.weight()) {
+                            // The weight of the new endpoint is greater than old one so we just remove the
+                            // old endpoint. The new endpoint will be added to
+                            // newlyAddedEndpoints at the bottom of this method.
+                            i.remove();
+                        }
+                        break;
+                    }
+                }
+            }
+            if (!replacedOldEndpoints.isEmpty()) {
+                oldEndpoints.addAll(replacedOldEndpoints);
+            }
+
+            for (final Iterator<EndpointsInUpdatingEntry> i = endpointsInUpdatingEntries.iterator();
+                 i.hasNext();) {
+                final EndpointsInUpdatingEntry endpointsInUpdatingEntry = i.next();
+
+                final Deque<EndpointAndStep> endpointAndSteps = endpointsInUpdatingEntry.endpointAndSteps();
+                removeIfNotInNewEndpoints(endpointAndSteps, newEndpoints);
+                if (endpointAndSteps.isEmpty()) {
+                    // All endpointAndSteps are removed so remove the entry from the Deque.
+                    i.remove();
+                    final ScheduledFuture<?> scheduledFuture = endpointsInUpdatingEntry.scheduledFuture;
+                    if (scheduledFuture != null) {
+                        scheduledFuture.cancel(true);
+                    }
+                }
+            }
+
+            final Deque<EndpointAndStep> newlyAddedEndpoints = new ArrayDeque<>();
+            for (Endpoint newEndpoint : newEndpoints) {
+                // We don't have to compare the weight because the old endpoint whose host is same and weight
+                // is different is already removed.
+                if (oldEndpoints.contains(newEndpoint)) {
+                    continue;
+                }
+
+                // Two for loops are used but it's only the number of endpoints in updating because there are
+                // no duplicate endpoints. So no big deal.
+                boolean found = false;
+                for (EndpointsInUpdatingEntry endpointsInUpdatingEntry : endpointsInUpdatingEntries) {
+                    for (EndpointAndStep endpointAndStep : endpointsInUpdatingEntry.endpointAndSteps()) {
```

Comment:
Question: Isn't it better to make `endpointAndStep` a `Set`? Do we need to use `Deque`?

Reply:
It's just that there's much chance `endpointAndStep` is removed from the first while iterating so I wanted to use a circular queue. But there's also a chance that the `endpointAndStep` in the middle is removed. So let me use the `Set`. 😉